core/rawdb/accessors_chain.go

```diff
@@ -79,6 +79,33 @@ func ReadAllHashes(db ethdb.Iteratee, number uint64) []common.Hash {
 	return hashes
 }
 
+// ReadAllCanonicalHashes retrieves all canonical number and hash mappings at the
+// certain chain range. Note, this method should only used in pruned light client
+// otherwise the cost can be very expensive.
+func ReadAllCanonicalHashes(db ethdb.Iteratee, from uint64, to uint64) ([]uint64, []common.Hash) {
```

Comment:
This iterator implementation seems a bit suboptimal to me. If I call this for mainnet with all blocks present, for `from = 8000000` and `to = from + 1`, you will iterate over 8M entries until you find the starting item. I guess if the use case is light client pruning you can expect all initial keys to be missing, but it still seems suboptimal.

Wouldn't a better solution be to use: `start = headerHashKey(from); `end = headerHashKey(to)` and then simply iterate with `NewIteratorWithStart(start)` and terminate when `bytes.Compare(key, end) >= 0`? We could also swap out `NewIteratorWithStart` to `NewIteratorWithRange` to make this code even simpler and the iterator even more flexible.